omp/reorder/rcm_kernels.cpp

```diff
@@ -0,0 +1,907 @@
+    if (strategy == gko::reorder::starting_strategy::minimum_degree) {
+        ubfs(exec, num_vertices, row_ptrs, col_idxs, &levels[0], min_idx,
+             max_val);
+        return min_idx;
+    }
+
+    // Intermediate storage.
+    // We copy back to `levels` for the final pick.
+    vector<IndexType> levels_clone(levels, exec);
+    vector<IndexType> current_levels_clone(levels, exec);
+
+    auto current = min_idx;
+    const auto contender_and_height =
+        rls_contender_and_height(exec, num_vertices, row_ptrs, col_idxs,
+                                 degrees, &levels_clone[0], current, max_val);
+    current_levels_clone.swap(levels_clone);
+    auto current_contender = contender_and_height.first;
+    auto current_height = contender_and_height.second;
+
+    // This loop always terminates, as height needs to strictly increase.
+    while (true) {
+        const auto contender_contender_and_height = rls_contender_and_height(
+            exec, num_vertices, row_ptrs, col_idxs, degrees, &levels_clone[0],
+            current_contender, max_val);
+        auto contender_contender = contender_contender_and_height.first;
+        auto contender_height = contender_contender_and_height.second;
+
+        if (contender_height > current_height) {
+            current_height = contender_height;
+            current = current_contender;
+            current_contender = contender_contender;
+            current_levels_clone = vector<IndexType>(levels, exec);
+            current_levels_clone.swap(levels_clone);
+        } else {
+            std::copy_n(current_levels_clone.begin(), num_vertices,
+                        levels.begin());
+            return current;
+        }
+    }
+}
+
+
+/**
+ * Counts how many nodes there are per level.
+ */
+template <typename IndexType>
+vector<IndexType> count_levels(std::shared_ptr<const OmpExecutor> exec,
+                               const IndexType *const levels,
+                               uint8 *const previous_component,
+                               size_type num_vertices)
+{
+    const auto num_threads = omp_get_max_threads();
+
+    // This initialization is necessary due to nested custom allocator.
+    vector<vector<IndexType>> level_counts(exec);
+    for (auto i = 0; i < num_threads; ++i) {
+        level_counts.push_back(vector<IndexType>(0, exec));
+    }
+
+#pragma omp parallel num_threads(num_threads)
+    {
+        const auto tid = omp_get_thread_num();
+        auto local_level_counts = &level_counts[tid];
+
+        auto current_size = 0;
+
+#pragma omp for schedule(static)
+        for (auto i = 0; i < num_vertices; ++i) {
+            const auto level = levels[i];
+
+            // If not part of previous component and actually discovered in this
+            // component.
+            if (!previous_component[i] &&
+                level != std::numeric_limits<IndexType>::max()) {
+                if (level >= current_size) {
```

Comment:
Nit: you could remove current_size completely
```suggestion
                if (level >= local_level_counts.size()) {
```